### PR TITLE
Feature 1642 action transaction

### PR DESCRIPTION
**Pull request by taojing2002**
See #1642 

Please double check two items in MetacatHandler.save:

- No accidental overwriting of the existing objects
- No accidental deleting of the existing objects

### CODE REVIEW

File: test/edu/ucsb/nceas/metacat/MetacatHandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,1793 @@
...
+import edu.ucsb.nceas.metacat.util.DocumentUtil;
+
+import static org.mockito.ArgumentMatchers.any;
+
+/**
+ * The test class of MetacatHandler
+ * @author tao
+ *
+ */
+public class MetacatHandlerTest {
```
Comment (artntek) [inline test/edu/ucsb/nceas/metacat/MetacatHandlerTest.java:63]: this is an integration test that relies on the database - can we rename it to `MetacatHandlerIT`?

File: test/edu/ucsb/nceas/metacat/MetacatHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,1793 @@
...
+        assertTrue("The object should have the checksum " + anotherChecksumStr + " rather than "
+                    + readChecksum, anotherChecksumStr.equals(readChecksum));
+    }
+
+    /**
+     * Test the save method to prevent overwriting an existing object
+     * @throws Exception
+     */
+    @Test
+    public void testSaveMethodWithMock() throws Exception {
```
Comment (artntek) [inline test/edu/ucsb/nceas/metacat/MetacatHandlerTest.java:1077]: The console output for this test method starts with:

```log
LeanTestUtils: 'test.printdebug' is TRUE
LeanTestUtils: PropertyService initialized in UNIT_TEST mode; using default properties from lib/metacat.properties overlaid with configurable properties from test/test.properties
LeanTestUtils: PropertyService initialized in LIVE_TEST mode; using default properties from /opt/local/share/java/tomcat/webapps/metacat/WEB-INF/metacat.properties overlaid with configurable properties from /var/metacat/config/metacat-site.properties
```

The call to initialize `UNIT_TEST` is set up in your `setUp()` method, but I can't see where you are making a call to initialize `LIVE_TEST` in this method. I'm guessing it is not intentional?

File: test/edu/ucsb/nceas/metacat/MetacatHandlerTest.java
[the remark this answers is not in the document]
Comment (artntek) [reply test/edu/ucsb/nceas/metacat/MetacatHandlerTest.java:1]: ### Comment 3
It's lovely to see such rigorous testing. On the other hand, it's really difficult to follow what is actually happening, when test methods are hundreds of lines long 😞 Consequently, I have not gone through all these at a fine-grained level. I'd encourage all of us to try to keep test methods very small for newly-created tests

File: test/edu/ucsb/nceas/metacat/dataone/CNodeServiceTest.java
Comment (artntek) [inline test/edu/ucsb/nceas/metacat/dataone/CNodeServiceTest.java:1]: rename to CNodeServiceIT?

File: test/edu/ucsb/nceas/metacat/dataone/MNodeServiceIT.java
Comment (artntek) [inline test/edu/ucsb/nceas/metacat/dataone/MNodeServiceIT.java:1]: see comments about `assertTrue` vs. `assertEquals` on MetacatHandlerTest - same applies to many of the ones added below

File: test/edu/ucsb/nceas/metacat/MetacatHandlerTest.java (hunk 2 above)
Comment (artntek) [reply test/edu/ucsb/nceas/metacat/MetacatHandlerTest.java:1077]: found it - line 1126. `D1NodeServiceTest` extends `MCTestCase`, which does this `LIVE_TEST` init.

File: test/edu/ucsb/nceas/metacat/systemmetadata/SystemMetadataValidatorTest.java [hunk 3]
```diff
@@ -9,35 +8,35 @@
...
-     * @param name  the name of test
-     */
-    public SystemMetadataValidatorTest(String name) {
-        super(name);
-    }
-    
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public class SystemMetadataValidatorTest {
```
Comment (artntek) [inline test/edu/ucsb/nceas/metacat/systemmetadata/SystemMetadataValidatorTest.java:21]: `SystemMetadataValidatorIT`?

File: test/edu/ucsb/nceas/metacattest/NonAsciiCharacterTest.java [hunk 4]
```diff
@@ -33,22 +8,33 @@
...
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
 /**
  * A JUnit test for testing Metacat when Non Ascii Characters are inserted
  */
-public class NonAsciiCharacterTest extends D1NodeServiceTest {
-    private Session session = null;
-    
+public class NonAsciiCharacterTest {
```
Comment (artntek) [inline test/edu/ucsb/nceas/metacattest/NonAsciiCharacterTest.java:32]: `NonAsciiCharacterIT`?

**Review by artntek — CHANGES_REQUESTED** [review]
Phew - that was intense 😅 

Please check all your tests and rename the ones that rely on database/metacat instance to `*IT`, if I missed any.

Thanks!

File: src/edu/ucsb/nceas/metacat/MetacatHandler.java [hunk 5]
```diff
@@ -113,317 +215,684 @@ public static InputStream read(String docid)
...
+    /**
+     * Save the object into disk
+     * @param sysmeta  the system metadata of the object, which contains the identifier
+     * @param changeModificationDate  if Metacat needs to update the modification date of the system
+     *              metadata when Metacat stores it.
+     * @param action  the action of the request: insert (new) or update
+     * @param docType  the type of object - data (BIN) or metadata
+     * @param object  the input stream contains the content of the object
+     * @param preSys  the system metadata of the obsoleted object in the update action.
+     *                It will be ignore in the update action.
```
[the remark this answers is not in the document]
Comment (taojing2002) [reply src/edu/ucsb/nceas/metacat/MetacatHandler.java:264]: It should be ignored when the request is an INSERT action.

File: src/edu/ucsb/nceas/metacat/MetacatHandler.java [hunk 6]
```diff
@@ -113,317 +215,684 @@ public static InputStream read(String docid)
...
+                            // exists except if the source and target are the same file, in
+                            // which case this method has no effect. So It should not overwrite an
+                            // existing data object.
+                            // https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html
+                            Files.move(tempFile.toPath(), newFile);
+                            long end = System.currentTimeMillis();
+                            logMetacat.info("Metacat only needs the move the object file from "
+                                    + "temporary location to the permanent location for the object "
+                                    + pid.getValue());
+                            logMetacat.info(edu.ucsb.nceas.metacat.common.Settings.PERFORMANCELOG
```
[the remark this answers is not in the document]
Comment (taojing2002) [reply src/edu/ucsb/nceas/metacat/MetacatHandler.java:583]: There were two Settings classes were imported. So I had to use the full name. But it is only one so we can import it.

File: test/edu/ucsb/nceas/metacat/dataone/MNodeServiceIT.java
Comment (taojing2002) [reply test/edu/ucsb/nceas/metacat/dataone/MNodeServiceIT.java:1]: I didn't trust the `assertEquals` for the string comparisons, so I used str.equal(str2).

File: src/edu/ucsb/nceas/metacat/DocumentImpl.java [hunk 7]
```diff
@@ -1158,367 +873,79 @@ private void getDocumentInfo(String docid, int revision) throws McdbException, E
...
-                    pstmt.close();
-                    conn.increaseUsageCount(1);
-
-                }
-            } else {
-                logMetacat.info("DocumentImp.delete - complete delete the document " + accnum);
-                pstmt = conn.prepareStatement(
-                    "SELECT * FROM xml_documents WHERE docid = ? and rev = ?");
+            logMetacat.info("DocumentImp.delete - complete delete the document " + accnum);
+            String query = "SELECT * FROM xml_documents WHERE docid = ? and rev = ?";
```
[the remark this answers is not in the document]
Comment (taojing2002) [reply src/edu/ucsb/nceas/metacat/DocumentImpl.java]: Now it selects the doctype from here and also the method of `getDocTypeFromDB()` was deleted.

**Review by artntek — COMMENTED** [review]
have reviewed all changes up to 8832504727beb32d51267f7df338da6591d7cd8a

File: src/edu/ucsb/nceas/metacat/dataone/D1NodeService.java [hunk 8]
```diff
@@ -2262,7 +1567,6 @@ protected Identifier archiveObject(
         throws InvalidToken, ServiceFailure, NotAuthorized, NotFound, NotImplemented {
```
[the remark this answers is not in the document]
Comment (taojing2002) [reply src/edu/ucsb/nceas/metacat/dataone/D1NodeService.java:1565]: The param does need. So I added the parameter in the `DocumentImpl.archive` method as well.

**Review by artntek — APPROVED** [review]

**Review by artntek — APPROVED** [review]

**Pull request merged**